pkg/joinserver/joinserver.go

```diff
@@ -122,3 +133,413 @@ func (js *JoinServer) RegisterHandlers(s *runtime.ServeMux, conn *grpc.ClientCon
+						"session_key_id",
+						"f_nwk_s_int_key",
+						"s_nwk_s_int_key",
+						"nwk_s_enc_key",
+						"app_s_key",
+					}, nil
+				},
+			)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			dev.Session = &ttnpb.Session{
+				StartedAt:   time.Now().UTC(),
+				DevAddr:     req.DevAddr,
+				SessionKeys: res.SessionKeys,
+			}
+			paths = append(paths, "session")
+
+			return dev, paths, nil
+		})
+	if err != nil {
+		logger.WithFields(log.Fields(
+			"join_eui", pld.JoinEUI,
+			"dev_eui", pld.DevEUI,
+		)).WithError(err).Error("Failed to update device")
+		return nil, err
+	}
+
+	registerAcceptJoin(ctx, dev, req)
+	return res, nil
+}
+
+// GetNwkSKeys returns the requested network session keys.
+func (js *JoinServer) GetNwkSKeys(ctx context.Context, req *ttnpb.SessionKeyRequest) (*ttnpb.NwkSKeysResponse, error) {
+	if dn, ok := auth.X509DNFromContext(ctx); ok {
+		dev, err := js.devices.GetByEUI(ctx, req.JoinEUI, req.DevEUI,
+			[]string{
+				"network_server_address",
+			},
+		)
+		if err != nil {
+			return nil, errRegistryOperation.WithCause(err)
+		}
+		if err := validateCaller(dn, dev.NetworkServerAddress); err != nil {
+			return nil, err
+		}
+	} else if err := clusterauth.Authorized(ctx); err != nil {
+		return nil, err
+	}
+
+	ks, err := js.keys.GetByID(ctx, req.DevEUI, req.SessionKeyID,
+		[]string{
+			"f_nwk_s_int_key",
+			"nwk_s_enc_key",
+			"s_nwk_s_int_key",
+		},
+	)
+	if err != nil {
+		return nil, errRegistryOperation.WithCause(err)
+	}
+
+	if ks.NwkSEncKey == nil {
+		return nil, errNoNwkSEncKey
+	}
+	if ks.FNwkSIntKey == nil {
+		return nil, errNoFNwkSIntKey
+	}
+	if ks.SNwkSIntKey == nil {
+		return nil, errNoSNwkSIntKey
+	}
+
+	return &ttnpb.NwkSKeysResponse{
+		NwkSEncKey:  *ks.NwkSEncKey,
+		FNwkSIntKey: *ks.FNwkSIntKey,
+		SNwkSIntKey: *ks.SNwkSIntKey,
+	}, nil
+}
+
+// GetAppSKey returns the requested application session key.
+func (js *JoinServer) GetAppSKey(ctx context.Context, req *ttnpb.SessionKeyRequest) (*ttnpb.AppSKeyResponse, error) {
+	if dn, ok := auth.X509DNFromContext(ctx); ok {
+		dev, err := js.devices.GetByEUI(ctx, req.JoinEUI, req.DevEUI,
+			[]string{
+				"application_server_address",
+			},
+		)
+		if err != nil {
+			return nil, errRegistryOperation.WithCause(err)
+		}
+		if err := validateCaller(dn, dev.ApplicationServerAddress); err != nil {
+			return nil, err
+		}
+	} else if err := clusterauth.Authorized(ctx); err != nil {
+		return nil, err
+	}
+
+	ks, err := js.keys.GetByID(ctx, req.DevEUI, req.SessionKeyID,
+		[]string{
+			"app_s_key",
+		},
+	)
+	if err != nil {
+		return nil, errRegistryOperation.WithCause(err)
+	}
+	if ks.AppSKey == nil {
+		return nil, errNoAppSKey
+	}
+	return &ttnpb.AppSKeyResponse{
+		AppSKey: *ks.AppSKey,
+	}, nil
+}
+
+// GetHomeNetID returns the requested NetID.
+func (js *JoinServer) GetHomeNetID(ctx context.Context, joinEUI, devEUI types.EUI64) (*types.NetID, error) {
+	if _, ok := auth.X509DNFromContext(ctx); !ok {
```

Comment:
Why check the presence of this, if the value is not validated?

Reply:
Also, please fix the comment at https://github.com/TheThingsNetwork/lorawan-stack/blob/48662cc8fb3201261c9d7ebc1b714ce337a38dae/pkg/auth/pkix.go#L31

Reply:
There is no origin validation for getting the Home NS in Backend Interfaces. This means that any Network Server that receives a join-request can request the Home NS on the Join Server as long as the calling NS is authenticated. This way, we simply check if there is authentication. If there's not, we require cluster auth.